nmoscommon/auth/auth_registrar.py

```diff
@@ -0,0 +1,200 @@
+            reg_resp = requests.post(registration_href,
+                                     data=data,
+                                     auth=('dannym', 'password'),
+                                     timeout=0.5,
+                                     proxies={'http': ''}
+                                     )
+            reg_resp.raise_for_status()  # Raise error if status !=201
+            return reg_resp.json()
+        except HTTPError as e:
+            print("Unable to Register Client with Auth Server. {}".format(e))
+
+    def initialise(self):
+        try:
+            # Check if credentials file already exists i.e. device is already registered
+            if os.path.isfile(CREDENTIALS_PATH):
+                print("Credentials file already exists. Using existing credentials.")
+                self.client_id, self.client_secret = get_credentials_from_file(CREDENTIALS_PATH)
+            else:
+                print("Registering with Authorization Server...")
+                reg_resp_json = self.send_registration_request()
+                self.write_credentials_to_file(reg_resp_json, CREDENTIALS_PATH)
+            return True
+        except Exception as e:
+            print("Unable to initialise OAuth Client with client credentials. {}".format(e))
+            return False
+
+
+class AuthRequestor(object):
+
+    def __init__(self, auth_registrar=None):
+        self.auth_reg = auth_registrar
+        self.bridge = IppmDNSBridge()
+
+    def _auth_request(self, data):
+        try:
+            href = self.bridge.getHref(MDNS_SERVICE_TYPE)
+            if href == '':
+                raise ValueError("No Services Found of type {}".format(MDNS_SERVICE_TYPE))
+            token_href = urljoin(href, TOKEN_ENDPOINT)
+            client_id, client_secret = get_credentials_from_file(CREDENTIALS_PATH)
+
+            # print('Data is: {}. HREF is: {}. ID is: {}. Secret is: {}'.format(
+            #     data, token_href, client_id, client_secret
+            # ))  # Only for Testing
+
+            token_req = requests.post(token_href,
+                                      data=data,
+                                      auth=(client_id, client_secret),
+                                      timeout=0.5,
+                                      proxies={'http': ''})
+            token_req.raise_for_status()
+            print("Token Request Successful")
+            return token_req.json()
+        except Exception as e:
+            print("Bearer Token Request Failed. {}".format(e))
+            return False
+
+    def token_request_password(self, username, password, scope):
+        return self.token_request(grant_type="password", username=username, password=password, scope=scope)
+
+    def token_request_code(self, code, scope):
+        return self.token_request(grant_type="authorization_code", auth_code=code, scope=scope)
+
+    def token_request(self, grant_type, username=None, password=None, auth_code=None, scope=None):
```

Comment:
I kinda feel this method is getting a bit too big. Could you look at breaking it down a bit please? I know most of it is just object defs using up lines, but I think there's scope to reduce it down a bit even so.

Reply:
I'd like to kindly disagree on this one ... My reasoning is, I've designed it so the `_auth_request` function is a universal, grant-agnostic function for requesting tokens.. the `token_request` function is merely a utility function that filters the data to be passed to `_auth_request`, with individual grant-speicific functions (e.g. `token_request_password`) acting as the API. So this function is the bridge between the API and core functions. Plus, as you say, it only actually consists of a conditional statement that assigns a value to `data` (I'd use a case statement if Python supported it)